--- a/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64-rootdisk.dts	2022-09-11 20:36:04.461349640 +0000
+++ b/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64-rootdisk.dts	2022-09-11 20:35:49.949267833 +0000
@@ -22,7 +22,7 @@
 
 	chosen {
 		stdout-path = "serial0:115200n8";
-		bootargs = "earlycon=uart8250,mmio32,0x11002000 console=ttyS0,115200n1 swiotlb=512 root=/dev/mmcblk0p7 rootfstype=squashfs,f2fs";
+		bootargs = "console=ttyS0,115200n1 rootfstype=squashfs earlyprintk block2mtd.block2mtd=/dev/mmcblk0,65536,eMMC,5 mtdparts=eMMC:768k(preloader)ro,512k(uboot),512k(uboot-env),256k(Factory),64M(firmware),2048M(usrdata)";
 	};
 
 	cpus {
@@ -220,6 +220,49 @@
 	assigned-clocks = <&topckgen CLK_TOP_MSDC30_0_SEL>;
 	assigned-clock-parents = <&topckgen CLK_TOP_UNIV48M>;
 	non-removable;
+	emmc@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "mmc-card";
+		reg = <0>;
+
+		partitions {
+				compatible = "fixed-partitions";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				partition@0 {
+						label = "preloader";
+						reg = <0x0000000 0x000c0000>;
+						read-only;
+				};
+
+				partition@c0000 {
+						label = "uboot";
+						reg = <0x00c0000 0x00140000>;
+				};
+
+				partition@140000 {
+						label = "uboot-env";
+						reg = <0x0140000 0x001c0000>;
+				};
+
+				factory: partition@1c0000 {
+						label = "Factory";
+						reg = <0x01c0000 0x00200000>;
+				};
+
+				partition@200000 {
+						label = "firmware";
+						reg = <0x0200000 0x02200000>;
+				};
+
+				partition@2200000 {
+						label = "usrdata";
+						reg = <0x2200000 0x82200000>;
+				};
+		};
+	};
 };
 
 &mmc1 {
@@ -529,6 +572,10 @@
 	};
 };
 
+&wmac {
+	status = "okay";
+};
+
 &pwm {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pwm7_pins>;
@@ -543,11 +590,11 @@
 };
 
 &sata {
-	status = "disable";
+	status = "okay";
 };
 
 &sata_phy {
-	status = "disable";
+	status = "okay";
 };
 
 &spi0 {
